### PR TITLE
Setup script for Chill

### PR DESCRIPTION
**Pull request by pratishshr**
## Features
- Asks user for the inputs required by chill.
- Modifies `chill.yml` as per the inputs.
- Uses `pm2` to start the processes.
- Displays the addresses where the processes are run.

## Limitations
- Doesn't add an `nginx` configuration. User will have to manually point to the addresses displayed.
- `node`/`yarn`/`pm2`/`git` have to be pre-installed for the script to proceed with the installation. (The script however does check if these are pre-installed or not)

## Screenshots
image
image

### CODE REVIEW

File: scripts/setup.sh [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+  read -p "Project logo Url: " logoUrl;
+
+  while true; do
+    read -p "Enter the URL of the service you would like to monitor: " serviceUrl
+    read -p "Enter the Name of this service: " serviceName
+    serviceUrls+=($serviceUrl)
+    serviceNames+=($serviceName)
+
+    read -p "Would you like to add her service?: (y/n) " addNewService
```
Comment (romitamgai) [inline scripts/setup.sh]: `her service?` :wink:

**Comment by romitamgai** [comment]
:100:

File: scripts/setup.sh (hunk 1 above)
Comment (kabirbaidhya) [reply scripts/setup.sh]: another service probably?

File: package.json [hunk 2]
```diff
@@ -54,7 +54,7 @@
     "nodemailer": "^4.0.1",
     "request": "^2.72.0",
     "request-promise": "^4.1.0",
-    "sqlite3": "^3.1.9",
+    "sqlite3": "^4.0.4",
```
Comment (kabirbaidhya) [inline package.json]: @pratishshr Let's do this in another specific PR. This isn't related to the setup script.

File: scripts/setup.sh (hunk 1 above)
Comment (pratishshr) [reply scripts/setup.sh]: haha, yes "another" service.

File: scripts/setup.sh [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+  read -p "Name of the project (lowercased without any spaces): ($name) " name
+  name=${name:-$DEFAULT_NAME}
+
+  read -p "API Port: " apiPort
+  read -p "Websocket Port: " wsPort
+  read -p "Domain name: (eg. chill.lftechnology.com) " domainName
+  read -p "Would you like to enable Slack notifications?: (y/n) " enableNotification
+
+  if [ "$enableNotification" = "y" ]; 
+  then
```
Comment (kabirbaidhya) [inline scripts/setup.sh]: `then` could be inline with `if` statement - that's the general convention too.

File: scripts/setup.sh [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+  if [ "$enableNotification" = "y" ]; 
+  then 
+    slackEnabled=true;
+  else 
+    slackEnabled=false;
+  fi
+
+  if $slackEnabled;
+  then
```
Comment (kabirbaidhya) [inline scripts/setup.sh]: .

File: scripts/setup.sh [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+  while true; do
+    read -p "Enter the URL of the service you would like to monitor: " serviceUrl
+    read -p "Enter the Name of this service: " serviceName
+    serviceUrls+=($serviceUrl)
+    serviceNames+=($serviceName)
+
+    read -p "Would you like to add another service?: (y/n) " addNewService
+    if ! [ "$addNewService" = "y" ]; 
+    then
```
Comment (kabirbaidhya) [inline scripts/setup.sh]: .

File: scripts/setup.sh [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    fi
+  done
+}
+
+# Check programs required for chill installation.
+check_programs() {
+  for i in "${REQUIRED_PROGRAMS[@]}"
+  do
+    if ! has_program $i;
+    then
```
Comment (kabirbaidhya) [inline scripts/setup.sh]: .

File: scripts/setup.sh [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+# Run build scripts in all services.
+build_services() {
+  build $FOLDER_CHILL_MONITOR;
+  build $FOLDER_CHILL_REST_API;
+  build $FOLDER_CHILL_DASHBOARD;
+}
+
+# Start all the services.
+start_services() {
+ CHILL_CONFIG=$CHILL_CONFIG pm2 start $FOLDER_CHILL_MONITOR/dist/index.js --name $name-monitor;
```
Comment (kabirbaidhya) [inline scripts/setup.sh]: Could be better clean and intuitive here.
```bash
CHILL_CONFIG=$CHILL_CONFIG pm2 start "${$FOLDER_CHILL_MONITOR}/dist/index.js" --name "${name}-monitor";

File: scripts/setup.sh [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+ CHILL_CONFIG=$CHILL_CONFIG pm2 start $FOLDER_CHILL_REST_API/dist/index.js --name $name-rest-api;
```
Comment (kabirbaidhya) [inline scripts/setup.sh]: same here

File: scripts/setup.sh [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+slackEndpoint=""
+slackEnabled=false
+name=$DEFAULT_NAME
+version=$DEFAULT_VERSION
+
+
+##### Functions #####
+# Replace string in file.
+replace_string() {
+  sed -i '' "s|$1|$2|g" $3
```
Comment (mesaugat) [inline scripts/setup.sh]: Are we sure this runs in linux? There are differences in `GNU` and `BSD` command line tools.

**Comment by mesaugat** [comment]
@kabirbaidhya @romitamgai Someone with `Linux` should try doing this and then we can merge it.

**Comment by romitamgai** [comment]
Also the script creates chill inside chill #chill-ception. 
`/codehome/leapfrog/chill-project/chill/scripts/chill`, `/codehome/leapfrog/chill-project/chill/scripts/chill/chill.yml`
@leapfrogtechnology/chill-team

File: scripts/setup.sh [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+    if ! has_program $i; then
+      echo "chill: Please install '"$i"' to continue the installation process.";
+      exit 0;
+    fi
+  done
+}
+
+# Clone all the repositories required for Chill.
+clone_repositories() {
+  git clone $LINK_CHILL_MONITOR;
```
Comment (mesaugat) [inline scripts/setup.sh]: Might want to clone it in a `/tmp` location? Or might want to clean up everything as a final step.

Also, what if the directory already exists? Might want to remove the directories first.

**Comment by mesaugat** [comment]
@manisacharya You can look at this script and update it. We need a `sh` file to setup all of chill. This should be like a one click install but from the terminal.

Good opportunity to learn a bit of bash. 😛

**Comment by manisacharya** [comment]
@mesaugat I was looking into this. First I had to move `chill.yml` outside the monitor folder. Now, I will work on this.

**Pull request closed without merging**